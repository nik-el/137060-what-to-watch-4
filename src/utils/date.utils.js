const MONTHS = [
  `January`,
  `February`,
  `March`,
  `April`,
  `May`,
  `June`,
  `July`,
  `August`,
  `September`,
  `October`,
  `November`,
  `December`
];

export const getFormattedDate = (dateString) => {
  if (!dateString) {
    return ``;
  }

  const date = new Date(dateString);
  return `${MONTHS[date.getMonth()]} ${date.getDate()}, ${date.getFullYear()}`;
};
