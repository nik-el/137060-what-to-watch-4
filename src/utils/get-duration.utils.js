const HOUR_MIN = 60;
const HOUR_SEC = 3600;

export const geFormatDuration = (time) => {
  const hours = Math.floor(time / HOUR_MIN);
  const min = time % HOUR_MIN;

  return `${hours}h ${min}m`;
};

export const geVideoDuration = (time) => {
  const secNum = parseInt(time, 10);
  let hours = Math.floor(secNum / HOUR_SEC);
  let minutes = Math.floor((secNum - (hours * HOUR_SEC)) / HOUR_MIN);
  let seconds = secNum - (hours * HOUR_SEC) - (minutes * HOUR_MIN);

  if (hours < 10) {
    hours = `0` + hours;
  }
  if (minutes < 10) {
    minutes = `0` + minutes;
  }
  if (seconds < 10) {
    seconds = `0` + seconds;
  }
  return hours + `:` + minutes + `:` + seconds;
};
