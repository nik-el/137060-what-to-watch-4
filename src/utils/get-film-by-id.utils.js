export const getFilmById = (films, id) => {
  if (!Array.isArray(films) || !id) {
    return null;
  }

  return films.find((item) => item.id === id);
};
