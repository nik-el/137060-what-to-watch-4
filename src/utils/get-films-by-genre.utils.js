export const getFilmsByGenre = (films, genre) => {
  if (!Array.isArray(films)) {
    return [];
  }
  if (!genre) {
    return films;
  }

  return films.filter((item) => item.genre === genre);
};
