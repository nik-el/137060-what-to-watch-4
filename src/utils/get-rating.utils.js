export const getRating = (rating) => {
  switch (true) {
    case (rating < 3):
      return `Bad`;
    case (rating >= 3 && rating < 5):
      return `Normal`;
    case (rating >= 5 && rating < 8):
      return `Good`;
    case (rating >= 8 && rating < 10):
      return `Very good`;
    case (rating === 10):
      return `Awesome`;
    default:
      return `No rating`;
  }
};
