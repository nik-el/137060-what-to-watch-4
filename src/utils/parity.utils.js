const getOddArray = (array) => {
  if (!Array.isArray(array)) {
    return [];
  }
  return array.filter((item, index) => {
    return index % 2;
  });

};
const getEvenArray = (array) => {
  if (!Array.isArray(array)) {
    return [];
  }

  return array.filter((item, index) => {
    return !(index % 2);
  });
};

export {
  getOddArray,
  getEvenArray
};
