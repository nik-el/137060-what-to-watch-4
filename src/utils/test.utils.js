const testFilm = {
  title: `Test title`,
  thumbnail: `https://picsum.photos/id/0`,
  id: `0`,
  genre: `Comedy`,
  poster: `https://picsum.photos/id/2`,
  description: `Test description`,
  rating: `5,5`,
  ratingCount: 100,
  directors: [`Test director`],
  starring: [`Test Star 1`, `Test Star 2`],
  preview: `http://test_preview.com`
};
Object.freeze(testFilm);

const testFilms = [
  {
    title: `Test title 1`,
    thumbnail: `https://picsum.photos/id/1`,
    id: `1`,
    genre: `test genre 1`,
    poster: `https://picsum.photos/id/2`,
    description: `Test description`,
    rating: `5,5`,
    ratingCount: 100,
    directors: [`Test director`],
    starring: [`Test Star 1`, `Test Star 2`]
  },
  {
    title: `Test title 2`,
    thumbnail: `https://picsum.photos/id/2`,
    id: `2`,
    genre: `test genre 2`,
    poster: `https://picsum.photos/id/2`,
    description: `Test description`,
    rating: `5,5`,
    ratingCount: 100,
    directors: [`Test director`],
    starring: [`Test Star 1`, `Test Star 2`]

  },
];
Object.freeze(testFilms);

const testGenres = [`test genre 1`, `test genre 2`];
Object.freeze(testGenres);

export {testFilm, testFilms, testGenres};
