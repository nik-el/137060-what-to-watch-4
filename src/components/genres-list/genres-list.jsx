import React from 'react';
import {useSelector, useDispatch} from "react-redux";

import {capitalize} from '../../utils/capitalize.utils';
import {ActionCreator} from "../../reducer";

const DEFAULT_GENRE = null;

export const GenresList = React.memo(function GenresList() {
  const dispatch = useDispatch();

  const handleGenreClick = (event, genre) => {
    event.preventDefault();
    dispatch(ActionCreator.setFilter(genre));
    dispatch(ActionCreator.getFilmsByGenre(genre));
  };

  const currentGenre = useSelector((state) => state.currentGenre);
  const genres = useSelector((state) => state.genres);

  const getGenreClass = (genre) => (
    `catalog__genres-item` + (currentGenre === genre ? ` catalog__genres-item--active` : ``)
  );

  return <ul className="catalog__genres-list">
    {
      [DEFAULT_GENRE, ...genres].map((genre, i)=> (
        <li
          key={genre + i}
          className={getGenreClass(genre)}
        >
          <a
            onClick={(e) => handleGenreClick(e, genre)}
            href="#"
            className="catalog__genres-link">{!genre ? `All genres` : capitalize(genre)}</a>
        </li>
      ))
    }
  </ul>;
});
