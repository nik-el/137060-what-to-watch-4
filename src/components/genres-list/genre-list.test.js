import React from 'react';
import renderer from 'react-test-renderer';
import {GenresList} from './genres-list';
import {BrowserRouter as Router} from 'react-router-dom';

import {testFilms, testGenres} from '../../utils/test.utils';

const onGenreClick = () => {};

it(`GenresList renders correctly`, () => {
  const tree = renderer
    .create(
        <Router>
          <GenresList
            films={testFilms}
            genres={testGenres}
            currentGenre={testGenres[0]}
            onGenreClick={onGenreClick}
          />
        </Router>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

