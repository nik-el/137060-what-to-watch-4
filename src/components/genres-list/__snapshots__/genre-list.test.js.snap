// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GenresList renders correctly 1`] = `
<ul
  className="catalog__genres-list"
>
  <li
    className="catalog__genres-item"
  >
    <a
      className="catalog__genres-link"
      href="#"
      onClick={[Function]}
    >
      All genres
    </a>
  </li>
  <li
    className="catalog__genres-item catalog__genres-item--active"
  >
    <a
      className="catalog__genres-link"
      href="#"
      onClick={[Function]}
    >
      Test genre 1
    </a>
  </li>
  <li
    className="catalog__genres-item"
  >
    <a
      className="catalog__genres-link"
      href="#"
      onClick={[Function]}
    >
      Test genre 2
    </a>
  </li>
</ul>
`;
